crates/c-api/src/wasi.rs

```diff
@@ -0,0 +1,274 @@
+            CStr::from_ptr(values[i]).to_bytes(),
+        );
+    }
+
+    std::mem::forget(config);
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn wasi_config_inherit_env(config: *mut wasi_config_t) {
+    let mut config = Box::from_raw(config);
+    config.builder = config.builder.inherit_env();
+    std::mem::forget(config);
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn wasi_config_set_stdin(
+    config: *mut wasi_config_t,
+    path: *const c_char,
+) -> bool {
+    let file = match open_file(path) {
+        Some(f) => f,
+        None => return false,
+    };
+
+    let mut config = Box::from_raw(config);
+    config.builder = config.builder.stdin(file);
+    std::mem::forget(config);
+
+    true
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn wasi_config_inherit_stdin(config: *mut wasi_config_t) {
+    let mut config = Box::from_raw(config);
+    config.builder = config.builder.inherit_stdin();
+    std::mem::forget(config);
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn wasi_config_set_stdout(
+    config: *mut wasi_config_t,
+    path: *const c_char,
+) -> bool {
+    let file = match create_file(path) {
+        Some(f) => f,
+        None => return false,
+    };
+
+    let mut config = Box::from_raw(config);
+    config.builder = config.builder.stdout(file);
+    std::mem::forget(config);
+
+    true
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn wasi_config_inherit_stdout(config: *mut wasi_config_t) {
+    let mut config = Box::from_raw(config);
+    config.builder = config.builder.inherit_stdout();
+    std::mem::forget(config);
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn wasi_config_set_stderr(
+    config: *mut wasi_config_t,
+    path: *const c_char,
+) -> bool {
+    let file = match create_file(path) {
+        Some(f) => f,
+        None => return false,
+    };
+
+    let mut config = Box::from_raw(config);
+    config.builder = config.builder.stderr(file);
+    std::mem::forget(config);
+
+    true
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn wasi_config_inherit_stderr(config: *mut wasi_config_t) {
+    let mut config = Box::from_raw(config);
+    config.builder = config.builder.inherit_stderr();
+    std::mem::forget(config);
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn wasi_config_preopen_dir(
+    config: *mut wasi_config_t,
+    path: *const c_char,
+    guest_path: *const c_char,
+) -> bool {
+    let guest_path = match cstr_to_path(guest_path) {
+        Some(p) => p,
+        None => return false,
+    };
+
+    let dir = match cstr_to_path(path) {
+        Some(p) => match preopen_dir(p) {
+            Ok(d) => d,
+            Err(_) => return false,
+        },
+        None => return false,
+    };
+
+    let mut config = Box::from_raw(config);
+    config.builder = config.builder.preopened_dir(dir, guest_path);
+    std::mem::forget(config);
+
+    true
+}
+
+#[repr(C)]
+pub struct wasi_instance_t {
+    wasi: Wasi,
+    export_cache: HashMap<String, *mut wasm_extern_t>,
+}
+
+impl Drop for wasi_instance_t {
```

Comment:
Instead of using a destructor here, could the values be stored as `Box<wasm_extern_t>`?

Reply:
Unfortunately not as that would require consuming the box to return a cached pointer for `wasi_instance_bind_import`, would it not?

Reply:
Oh I think you could update `wasi_instance_bind_import` by basically converting from` &Box<wasm_extern_t>` (from the lookup in the hash map) to `&wasm_extern_t` to `*const wasm_extern_t`

Reply:
I assume `Box<T>` layout guarantees that cast?

Reply:
Fixed.